lib/espresso-runner.js

```diff
@@ -65,46 +64,30 @@ class EspressoRunner {
     return !signed;
   }
 
-  // TODO duplicated from uiautomator2, should be a lib method
-  async getAPKVersion (apk) {
-    let args = ['dump', 'badging', apk];
-    await this.adb.initAapt();
-    let {stdout} = await exec(this.adb.binaries.aapt, args);
-    let apkVersion = new RegExp(/versionName='([^']+)'/g).exec(stdout);
-    if (apkVersion && apkVersion.length >= 2) {
-      apkVersion = apkVersion[1];
-    } else {
-      apkVersion = null;
-    }
-    return apkVersion.toString();
-  }
-
-  // TODO duplicated from uiautomator2, should be a lib method
-  async getInstalledPackageVersion (pkg) {
-    let stdout =  await this.adb.shell(['dumpsys', 'package', pkg]);
-    let pkgVersion = new RegExp(/versionName=([^\s\s]+)/g).exec(stdout);
-    if (pkgVersion && pkgVersion.length >= 2) {
-      pkgVersion = pkgVersion[1];
-    } else {
-      pkgVersion = null;
-    }
-    return pkgVersion.toString();
-  }
-
   async startSession (caps) {
     let cmd = ['am', 'instrument', '-w', '-e', 'debug', 'false',
       `${TEST_APK_PKG}/android.support.test.runner.AndroidJUnitRunner`];
 
     logger.info(`Starting Espresso Server /*TODO VERSION*/ with cmd: ${cmd.join(' ')}`);
 
-    await this.adb.shell(cmd);
+    // start the instrumentation process, but do not wait for it
+    // otherwise it will block for longer than needed for the device to be ready
+    let err;
+    this.adb.shell(cmd).catch((e) => { // eslint-disable-line
+      // handle asynchronous errors that no longer get thrown
+      err = e;
+    });
 
     logger.info('Waiting for Espresso to be online...');
+
     // wait 20s for espresso to be online
     await retryInterval(20, 1000, async () => {
+      if (err) return; // eslint-disable-line curly
       await this.jwproxy.command('/status', 'GET');
     });
+    if (err) throw err; // eslint-disable-line curly
```

Comment:
it's probably clear here, that instruments have been successfully started, so it should be safer to await it here in comparison to just checking the result of err:

```
let err;
let instrumentsExecutor =  this.adb.shell(cmd).catch((e) => { // eslint-disable-line
      // handle asynchronous errors that no longer get thrown
      err = e;
    });
...
await instrumentsExecutor;
```